util/importer.py

```diff
@@ -8,7 +8,14 @@
 from django.conf import settings
 
 
-def import_path(path: str, module_name: str = "", force_reload = False):
+# this is needed so that import from course files works correctly.
+# it seems like importing a module requires the whole dotted path to a
+# path in sys.path, otherwise imports within the module's package fail
```

Comment:
```suggestion
# This is needed for importing modules defined in the course repo.
# For example, the exercise `view_type` may refer to Python modules in the course repo.
# The path to course-specific view types is given relative to the course directory.
# For example, `view_type: .exercises.jsav.views.iframe.view`
```
I think the comment was not very easy to understand. Is my version correct?

By the way, if the Python code runs `import exercises.jsav.views.iframe`, then doesn't it make a lot of sense that PYTHONPATH must include the course root directory (because the exercises subdirectory starts from there)? If PYTHONPATH contained the jsav directory, then it would NOT find package `exercises` there.

The `import_named` function actually adds the course key (directory) to the path before importing, thus the import statement to run is like `import traky.exercises.jsav.views.iframe`. Therefore, PYTHONPATH must contain `settings.COURSES_PATH`.

But does the MOOC-Grader now contain subdirectories `publish` and `store` under `COURSES_PATH` so that the course directories are not directly under `COURSES_PATH`?

Reply:
So, this is a bit complicated. We actually can import and run the exercises.jsav.views.iframe file using the code that we already have, as it is importing the file directly through its file location and not doing 'import exercises.jsav.views.iframe'. The problems arise when the iframe module itself tries to import something from the course directory. E.g. if iframe.py tries to do 'from .test import MyClass', we would get an error 'traky is not a module', or if it tries to do 'from views.test import MyClass', we'd get 'views is not a package'. My understanding is that this happens because it cannot find the traky folder in sys.path, so it cannot import it (and therefore its descendants).

> By the way, if the Python code runs import exercises.jsav.views.iframe, then doesn't it make a lot of sense that PYTHONPATH must include the course root directory (because the exercises subdirectory starts from there)? If PYTHONPATH contained the jsav directory, then it would NOT find package exercises there.

See above for comment about ' import exercises.jsav.views.iframe'.

Yes, it wouldn't find exercises package if the directory above it isn't in sys.path or the package above exercises hasn't been loaded yet. We shouldn't need the course directory itself in sys.path as we are including it in the import path when we load the view (e.g. traky.exercises.jsav.views.iframe)

> The import_named function actually adds the course key (directory) to the path before importing, thus the import statement to run is like import traky.exercises.jsav.views.iframe. Therefore, PYTHONPATH must contain settings.COURSES_PATH.

See above for the explanation as to why COURSES_PATH is need in sys.path.

> But does the MOOC-Grader now contain subdirectories publish and store under COURSES_PATH so that the course directories are not directly under COURSES_PATH?

COURSES_PATH is the publish/ directory.

Reply:
Thanks for the explanation!

> COURSES_PATH is the publish/ directory.

Not in the default `grader/settings.py`, though? Shouldn't that be fixed?

You have force-pushed a couple of times, but your branch actually isn't rebased onto the upstream master. You could that now.

Reply:
In the default settings courses/ is the publish directory, and course_store/ is the store directory. The publish directory is courses/publish/ only in the Aalto production server

Reply:
> In the default settings courses/ is the publish directory, and course_store/ is the store directory. The publish directory is courses/publish/ only in the Aalto production server

OK. Why was it done this way? Network file system?

Reply:
The publish and store directories need to be on the same filesystem for the os.rename to work. We thus need the store and publish directories to be under the same directory. Each directory under the project root is mounted separately to allow separation of config from other files, so they are both under the courses directory.